scripts/Diffraction/isis_powder/pearl.py

```diff
@@ -89,9 +89,46 @@ def create_cal(self, **kwargs):
     def should_subtract_empty_inst(self):
         return self._inst_settings.subtract_empty_inst
 
+    def _generate_out_file_paths(self, run_details):
+        """
+        Generates the various output paths and file names to be used during saving or as workspace names
+        :param run_details: The run details associated with this run
+        :return: A dictionary containing the various output paths and generated output name
+        """
+        output_directory = os.path.join(self._output_dir, run_details.label, self._user_name)
+        output_directory = os.path.abspath(os.path.expanduser(output_directory))
+        xye_files_directory = output_directory
+        if self._inst_settings.dat_files_directory:
+            xye_files_directory = os.path.join(output_directory, self._inst_settings.dat_files_directory)
+
+        file_type = "" if run_details.file_extension is None else run_details.file_extension.lstrip(".")
+        out_file_names = {"output_folder": output_directory}
+        format_options = {
+            "inst": self._inst_prefix,
+            "instlow": self._inst_prefix.lower(),
+            "instshort": self._inst_prefix_short,
+            "runno": run_details.output_run_string,
+            "fileext": file_type,
+            "_fileext": "_" + file_type if file_type else "",
+            "suffix": run_details.output_suffix if run_details.output_suffix else "",
+        }
+        format_options = self._add_formatting_options(format_options)
+
+        output_formats = {
+            "nxs_filename": output_directory,
+            "gss_filename": os.path.join(output_directory, "GSAS"),
+            "tof_xye_filename": os.path.join(xye_files_directory, "ToF"),
+            "dspacing_xye_filename": os.path.join(xye_files_directory, "dSpacing"),
+        }
```

Comment:
This Pearl-specific function override of `_generate_out_file_paths` copies all of the code from the equivalent function in abstract_inst.py but I think the only changes are L119-121. The `dat_files_directory` variable is also renamed higher up but that's not important. Please could you create a new function in abstract_inst.py called `_get_output_formats()` or similar that generates the output_formats dictionary
and then have an overridden version of `_get_output_formats` in this unit that does the Pearl-specific behaviour (subfolders)? You can then remove the Pearl specific version of `_generate_out_file_paths` I think